In [3]:
## importing Important modules
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier,AdaBoostClassifier,GradientBoostingClassifier
from imblearn.over_sampling import SMOTE
import warnings
warnings.filterwarnings("ignore")
from sklearn.model_selection import RandomizedSearchCV
from xgboost import XGBClassifier

In [4]:
print("Loading TF-IDF features...")
df = pd.read_csv("../Datasets/mbti_tfidf_features.csv")

Loading TF-IDF features...


In [5]:
models={
    "Logistic Regression" : LogisticRegression(n_jobs=-1),
    "Random Forest"      : RandomForestClassifier(n_jobs=-1),
    "AdaBoost"  : AdaBoostClassifier(),
    "XGBoost"            : XGBClassifier(n_jobs=-1),
    "SVM"                 : SVC()
}

In [13]:
params={
    "Logistic Regression" : {
                                'C': [0.01, 0.1, 1, 10, 100],
                                'penalty': ['l2'],
                                'solver': ['lbfgs']
                            },
                            
    "Random Forest"         : {
                                'n_estimators': [100, 200, 300],
                                'max_depth': [5, 10, 20, None],
                                'min_samples_split': [2, 5, 10]
                            },

    "AdaBoost"            : {
                                
                                'learning_rate': [0.01]
                            },
    "XGBoost"             : {
                                'learning_rate': [0.01, 0.1, 0.2],
                                'max_depth': [3, 5, 7],
                                'n_estimators': [100, 200]
                            },

    "SVM"                 : {
                                'C': [1],
                                'kernel': ['linear'],
                            }

}

# Hyper Parameter Tunning for Intuition (N) – Sensing (S)

In [18]:
print("Intuition (N) - Sensing (S)")
X = df.drop(columns=['MBTI_type', 'IE', 'NS', 'TF', 'JP', 'original_posts', 'clean_posts'])
y = df['NS']
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)
smote = SMOTE(random_state=42)
X_train, y_train = smote.fit_resample(X_train, y_train)

for i in range(len(list(models))):
            model = list(models.values())[i]
            print(f"Prediction model : {list(models.keys())[i]}")
            para=params[list(models.keys())[i]]

            rs = RandomizedSearchCV(model,para,cv=3)
            rs.fit(X_train,y_train)

            model.set_params(**rs.best_params_)
            model.fit(X_train,y_train)

            y_pred = rs.predict(X_test)
            
            print("Classification Report:")
            print(classification_report(y_test, y_pred))
            print("Accuracy:", accuracy_score(y_test, y_pred))
            print('='*35)
            print('\n')

Intuition (N) - Sensing (S)
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.56      0.57       239
           1       0.93      0.93      0.93      1496

    accuracy                           0.88      1735
   macro avg       0.75      0.75      0.75      1735
weighted avg       0.88      0.88      0.88      1735

Accuracy: 0.8829971181556195


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.25      0.01      0.02       239
           1       0.86      1.00      0.92      1496

    accuracy                           0.86      1735
   macro avg       0.56      0.50      0.47      1735
weighted avg       0.78      0.86      0.80      1735

Accuracy: 0.8599423631123919


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.29      0.4

KeyboardInterrupt: 

Intuition (N) – Sensing (S)
Prediction model : Logistic Regression

Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.56      0.57       239
           1       0.93      0.93      0.93      1496

    accuracy                           0.88      1735
   macro avg       0.75      0.75      0.75      1735
weighted avg       0.88      0.88      0.88      1735

Accuracy: 0.8829971181556195
===================================


Prediction model : Random Forest
Classification Report:

              precision    recall  f1-score   support

           0       0.40      0.01      0.02       239
           1       0.86      1.00      0.93      1496

    accuracy                           0.86      1735

   macro avg       0.63      0.50      0.47      1735

weighted avg       0.80      0.86      0.80      1735

Accuracy: 0.861671469740634
===================================


Prediction model : AdaBoost

Classification Report:

              precision    recall  f1-score   support

           0       0.29      0.41      0.34       239
           1       0.90      0.84      0.87      1496

    accuracy                           0.78      1735

   macro avg       0.59      0.62      0.60      1735

weighted avg       0.81      0.78      0.80      1735

Accuracy: 0.7798270893371758
===================================


Prediction model : XGBoost

Classification Report:

              precision    recall  f1-score   support

           0       0.72      0.38      0.50       239
           1       0.91      0.98      0.94      1496

    accuracy                           0.89      1735

   macro avg       0.81      0.68      0.72      1735

weighted avg       0.88      0.89      0.88      1735

Accuracy: 0.8945244956772335
===================================

Prediction model : SVM

Classification Report:

              precision    recall  f1-score   support

           0       0.56      0.63      0.59       239
           1       0.94      0.92      0.93      1496

    accuracy                           0.88      1735

   macro avg       0.75      0.77      0.76      1735

weighted avg       0.89      0.88      0.88      1735


Accuracy: 0.8812680115273775
===================================

# Hyper Parameter Tunning for Introversion (I) – Extroversion (E)

In [14]:
print("Introversion (I) – Extroversion (E)")
X = df.drop(columns=['MBTI_type', 'IE', 'NS', 'TF', 'JP', 'original_posts', 'clean_posts'])
y = df['IE']
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)
smote = SMOTE(random_state=42)
X_train, y_train = smote.fit_resample(X_train, y_train)
for i in range(len(list(models))):
            model = list(models.values())[i]
            print(f"Prediction model : {list(models.keys())[i]}")
            para=params[list(models.keys())[i]]

            rs = RandomizedSearchCV(model,para,cv=3)
            rs.fit(X_train,y_train)

            model.set_params(**rs.best_params_)
            model.fit(X_train,y_train)

            y_pred = rs.predict(X_test)
            
            print("Classification Report:")
            print(classification_report(y_test, y_pred))
            print("Accuracy:", accuracy_score(y_test, y_pred))
            print('='*35)
            print('\n')

Introversion (I) – Extroversion (E)
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.59      0.59       400
           1       0.88      0.87      0.88      1335

    accuracy                           0.81      1735
   macro avg       0.73      0.73      0.73      1735
weighted avg       0.81      0.81      0.81      1735

Accuracy: 0.8080691642651296


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.09      0.16       400
           1       0.78      0.98      0.87      1335

    accuracy                           0.78      1735
   macro avg       0.68      0.54      0.51      1735
weighted avg       0.74      0.78      0.71      1735

Accuracy: 0.7752161383285303


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.49 

Introversion (I) – Extroversion (E)
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.59      0.59       400
           1       0.88      0.87      0.88      1335

    accuracy                           0.81      1735
   macro avg       0.73      0.73      0.73      1735
weighted avg       0.81      0.81      0.81      1735

Accuracy: 0.8080691642651296
===================================


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.58      0.09      0.16       400
           1       0.78      0.98      0.87      1335

    accuracy                           0.78      1735
   macro avg       0.68      0.54      0.51      1735
weighted avg       0.74      0.78      0.71      1735

Accuracy: 0.7752161383285303
===================================


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.49      0.46      0.48       400
           1       0.84      0.86      0.85      1335

    accuracy                           0.77      1735
   macro avg       0.67      0.66      0.66      1735
weighted avg       0.76      0.77      0.76      1735

Accuracy: 0.7659942363112392
===================================


Prediction model : XGBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.74      0.53      0.62       400
           1       0.87      0.94      0.91      1335

    accuracy                           0.85      1735
   macro avg       0.80      0.74      0.76      1735
weighted avg       0.84      0.85      0.84      1735

Accuracy: 0.8484149855907781
===================================


Prediction model : SVM
Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.67      0.66       400
           1       0.90      0.89      0.89      1335

    accuracy                           0.84      1735
   macro avg       0.77      0.78      0.78      1735
weighted avg       0.84      0.84      0.84      1735

Accuracy: 0.839193083573487
===================================

## Hyper Parameter Tunning for  Judging (J) – Perceiving (P)

In [16]:
print("Judging (J) – Perceiving (P)")
X = df.drop(columns=['MBTI_type', 'IE', 'NS', 'TF', 'JP', 'original_posts', 'clean_posts'])
y = df['JP']
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)
smote = SMOTE(random_state=42) 
X_train, y_train = smote.fit_resample(X_train, y_train)
for i in range(len(list(models))):
            model = list(models.values())[i]
            print(f"Prediction model : {list(models.keys())[i]}")
            para=params[list(models.keys())[i]]

            rs = RandomizedSearchCV(model,para,cv=3)
            rs.fit(X_train,y_train)

            model.set_params(**rs.best_params_)
            model.fit(X_train,y_train)

            y_pred = rs.predict(X_test)
            
            print("Classification Report:")
            print(classification_report(y_test, y_pred))
            print("Accuracy:", accuracy_score(y_test, y_pred))
            print('='*35)
            print('\n')

Judging (J) – Perceiving (P)
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.85      0.84      1048
           1       0.76      0.74      0.75       687

    accuracy                           0.81      1735
   macro avg       0.80      0.79      0.80      1735
weighted avg       0.80      0.81      0.80      1735

Accuracy: 0.8051873198847262


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.91      0.79      1048
           1       0.75      0.42      0.54       687

    accuracy                           0.72      1735
   macro avg       0.73      0.67      0.67      1735
weighted avg       0.72      0.72      0.69      1735

Accuracy: 0.7164265129682997


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.

Judging (J) – Perceiving (P)
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.85      0.84      1048
           1       0.76      0.74      0.75       687

    accuracy                           0.81      1735
   macro avg       0.80      0.79      0.80      1735
weighted avg       0.80      0.81      0.80      1735

Accuracy: 0.8051873198847262
===================================


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.71      0.91      0.79      1048
           1       0.75      0.42      0.54       687

    accuracy                           0.72      1735
   macro avg       0.73      0.67      0.67      1735
weighted avg       0.72      0.72      0.69      1735

Accuracy: 0.7164265129682997
===================================


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.88      0.76      1048
           1       0.65      0.33      0.43       687

    accuracy                           0.66      1735
   macro avg       0.66      0.60      0.60      1735
weighted avg       0.66      0.66      0.63      1735

Accuracy: 0.6628242074927954
===================================


Prediction model : XGBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.88      0.84      1048
           1       0.79      0.69      0.73       687

    accuracy                           0.80      1735
   macro avg       0.80      0.78      0.79      1735
weighted avg       0.80      0.80      0.80      1735

Accuracy: 0.801729106628242
===================================


Prediction model : SVM
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.83      0.83      1048
           1       0.74      0.72      0.73       687

    accuracy                           0.79      1735
   macro avg       0.78      0.78      0.78      1735
weighted avg       0.79      0.79      0.79      1735

Accuracy: 0.7878962536023055
===================================



## Hyper Parameter Tunning for Thinking (T) – Feeling (F) 

In [17]:
print("Thinking (T) – Feeling (F) ")
X = df.drop(columns=['MBTI_type', 'IE', 'NS', 'TF', 'JP', 'original_posts', 'clean_posts'])
y = df['TF']
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)
smote = SMOTE(random_state=42)
X_train, y_train = smote.fit_resample(X_train, y_train)
for i in range(len(list(models))):
            model = list(models.values())[i]
            print(f"Prediction model : {list(models.keys())[i]}")
            para=params[list(models.keys())[i]]

            rs = RandomizedSearchCV(model,para,cv=3)
            rs.fit(X_train,y_train)

            model.set_params(**rs.best_params_)
            model.fit(X_train,y_train)

            y_pred = rs.predict(X_test)
            
            print("Classification Report:")
            print(classification_report(y_test, y_pred))
            print("Accuracy:", accuracy_score(y_test, y_pred))
            print('='*35)
            print('\n')

Thinking (T) – Feeling (F) 
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.84      0.85       939
           1       0.82      0.85      0.83       796

    accuracy                           0.84      1735
   macro avg       0.84      0.84      0.84      1735
weighted avg       0.84      0.84      0.84      1735

Accuracy: 0.8420749279538905


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.86      0.84       939
           1       0.82      0.78      0.80       796

    accuracy                           0.82      1735
   macro avg       0.82      0.82      0.82      1735
weighted avg       0.82      0.82      0.82      1735

Accuracy: 0.8219020172910663


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.3

Thinking (T) – Feeling (F) 
Prediction model : Logistic Regression
Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.84      0.85       939
           1       0.82      0.85      0.83       796

    accuracy                           0.84      1735
   macro avg       0.84      0.84      0.84      1735
weighted avg       0.84      0.84      0.84      1735

Accuracy: 0.8420749279538905
===================================


Prediction model : Random Forest
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.86      0.84       939
           1       0.82      0.78      0.80       796

    accuracy                           0.82      1735
   macro avg       0.82      0.82      0.82      1735
weighted avg       0.82      0.82      0.82      1735

Accuracy: 0.8219020172910663
===================================


Prediction model : AdaBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.39      0.52       939
           1       0.55      0.89      0.68       796

    accuracy                           0.62      1735
   macro avg       0.68      0.64      0.60      1735
weighted avg       0.69      0.62      0.60      1735

Accuracy: 0.6178674351585014
===================================


Prediction model : XGBoost
Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.83      0.84       939
           1       0.81      0.82      0.81       796

    accuracy                           0.83      1735
   macro avg       0.83      0.83      0.83      1735
weighted avg       0.83      0.83      0.83      1735

Accuracy: 0.8270893371757925
===================================


Prediction model : SVM
Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.85      0.85       939
           1       0.82      0.83      0.83       796

    accuracy                           0.84      1735
   macro avg       0.84      0.84      0.84      1735
weighted avg       0.84      0.84      0.84      1735

Accuracy: 0.839193083573487
===================================

